### PR TITLE
[Impeller] scales blur coverage to match rendered output

### PR DESCRIPTION
**Pull request by gaaclarke**
in pursuit of: https://github.com/flutter/flutter/issues/131580

I'm starting to write tests for the new blur, but I wanted to make tests for the old blur to make sure I understood the coverage rules.  I decided to check this in separately while I work on the other blur.

Even though the goal is to delete this class eventually.  Most of this work transfers over to the new blur.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or the PR is [test-exempt]. See [testing the engine] for instructions on writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: impeller/entity/contents/filters/directional_gaussian_blur_filter_contents_unittests.cc [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+  contents->SetDirection({1.0, 0.0});
+  contents->SetInputs({FilterInput::Make(texture)});
+  ContentContext renderer(mock_context_, mock_typographer_context_,
+                          mock_render_target_allocator_);
+  Entity entity;
+  Rect coverage_hint = Rect::MakeLTRB(0, 0, 0, 0);
+  std::optional<Entity> result =
+      contents->GetEntity(renderer, entity, coverage_hint);
+  ASSERT_TRUE(result.has_value());
+  ASSERT_EQ(result.value().GetBlendMode(), BlendMode::kSourceOver);
```
Comment (gaaclarke) [inline impeller/entity/contents/filters/directional_gaussian_blur_filter_contents_unittests.cc:181]: I would like to have better asserts here about the coverage of the result.  I'm not 100% up to date about what the values should be considering the coverage hint yet though.

File: impeller/entity/contents/filters/directional_gaussian_blur_filter_contents_unittests.cc [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+  contents->SetSigma(Sigma{sigma_radius_1});
+  contents->SetDirection({1.0, 0.0});
+  std::shared_ptr<MockTexture> texture = std::make_shared<MockTexture>(desc);
+  EXPECT_CALL(*texture, GetSize()).WillRepeatedly(Return(ISize(100, 100)));
+  FilterInput::Vector inputs = {FilterInput::Make(texture)};
+  Entity entity;
+  entity.SetTransformation(Matrix::MakeTranslation({100, 100, 0}));
+  std::optional<Rect> coverage = contents->GetFilterCoverage(
+      inputs, entity, /*effect_transform=*/Matrix::MakeScale({2.0, 2.0, 1.0}));
+  ASSERT_EQ(coverage, Rect::MakeLTRB(100 - 2, 100, 200 + 2, 200));
```
Comment (matanlurey) [inline impeller/entity/contents/filters/directional_gaussian_blur_filter_contents_unittests.cc:129]: Here and elsewhere:

I might either expand on the test name, or leave a comment (or does ASSERT support reasons, i.e. << "Reason") of why this is the expected result. From someone totally unfamiliar with this code, I don't understand what the test would mean if it failed or why this is the expected result.

**Review by matanlurey — COMMENTED** [review]
Actual approval should come from someone who understands the implementation IMO

**Review by jonahwilliams — CHANGES_REQUESTED** [review]
I think this is an example I would use of questionable mock usage. the DirectionalGaussianBlur is written against the HAL, which we can already test on CI with the real implementations. See entity_unittests for examples, but we don't need to, nor should we mock out APIs that are already abstracted from the platform.

**Comment by matanlurey** [comment]
> I think this is an example I would use of questionable mock usage. the DirectionalGaussianBlur is written against the HAL, which we can already test on CI with the real implementations. See entity_unittests for examples, but we don't need to, nor should we mock out APIs that are already abstracted from the platform.

What is the value of the "real" HAL here? Or are you saying you wouldn't want to see tests that assert C++ math would want to see golden/UI tests produced from a real backend instead?

**Comment by jonahwilliams** [comment]
Assertions on the contents are great! but you don't need to mock out the HAL to write this test. Just ... use the HAL as it already exists?

**Review by jonahwilliams — COMMENTED** [review]
Basically to do so, you'd take a similar structure to the golden tests like https://github.com/flutter/engine/blob/main/impeller/entity/entity_unittests.cc#L261

Except rather than making an assertion on the pixels, you can introspect on the contents/entity

**Comment by gaaclarke** [comment]
> I think this is an example I would use of questionable mock usage. the DirectionalGaussianBlur is written against the HAL, which we can already test on CI with the real implementations. See entity_unittests for examples, but we don't need to, nor should we mock out APIs that are already abstracted from the platform.

@jonahwilliams I gave `entity_unittests` a look.  Having gone through the google accreditation process for writing c++ tests there are 2 guiding principles applicable here:  1) Everything has tests 2) Prefer concrete classes over mocks.

Preferring concrete classes is why in these tests I used a real `ContentContext` instead of just mocking that, for example.  It's also why I gave your suggestion serious consideration.  I mentioned it in the other PR but the approach you are proposing is flawed as-is because this layer of Impeller changes its behavior based on the capabilities of the hardware.  So without some effort it cannot be used for all testing needs.  One of my goals with this PR is to provide the tools so people can test any logic.

There are some potential scaling issues with that approach as well.  Managing hardware profiles can be more cumbersome than directly specifying them in the test with the mocks.  All tests will be run 3 times, which assuming the bottom layers are correct (a correct assumption for a unit test), will be redundant.  Plus they are incurring extra cost of starting and tearing down the hardware contexts.

I think it's really important we get the type of integration testing we'd get from using the real objects, for that we have the golden image tests.  I'm in the middle of writing out the new blur but I'm using this testing harness to make sure I get coverage correct.  I'll use golden images to make sure the visual results are correct.

Sound good?  I'm happy to discuss further over chat or brainjam on a design that broadens out your testing harness.

**Comment by gaaclarke** [comment]
Ugh, i really messed up this branch.  I accidentally had it tracking the branch that I tried to split this from.  I'll try to sort it out.

**Comment by jonahwilliams** [comment]
Why do you need to manage hardware profiles? It sounds like you're worried about test performance, but are we actually having problems with that, or is this a hypothetical concern? Its true that we have limited CI capacity, but that vast majority of that is getting burned on compilation, not test execution AFAIK (though @zanderso  can correct me).

**Comment by gaaclarke** [comment]
> Why do you need to manage hardware profiles? 

Imagine we had a system where we could have full test coverage using the real hardware contexts.  It would have to look something like:

```c++
TEST(Foo, bar) {
  std::string profile_path; 
  std::shared_ptr<Context> context;
  if (GetVariant() == "vulkan" || GetVariant() == "opengles") {
    profile_path = GetTextFixture("moto_g4.profile");
    context = MakeContext(profile_path);
  } else {
    context = MakeContext();
  }
  auto foo = Foo(context);
  ASSERT_TRUE(foo.Something());
}
```

Otherwise we'd have to make some way to to manipulate the contexts like this:
```c++
TEST(Foo, bar) {
  std::string profile_path; 
  std::shared_ptr<Context> context = MakeContext();
  context->DisableGPUFeatureX();
  auto foo = Foo(context);
  ASSERT_TRUE(foo.Something());
}
```

But that would only work for disabling features, not for enabling them.

>It sounds like you're worried about test performance

My main concern is test coverage.  I do have concerns about scaling in terms of management and test performance.  As someone that has worked on a project with large test coverage, it can make a difference and running the same test 3 times with hardware contexts is an order of magnitude slower.  If we were able to solve test coverage though we could consider overlooking that or finding some solution.

**Comment by jonahwilliams** [comment]
Where in this test do you need to customize the hardware profile?

**Comment by zanderso** [comment]
Sorry, not entirely following the discussion about hardware profiles, but in terms of CI cycles, except for running ASAN tests on debug_unopt builds, the tests on engine CI currently run quite fast, and it's correct that the slow step is C++ compilation.

**Comment by gaaclarke** [comment]
> Where in this test do you need to customize the hardware profile?

This test you don't.  But if you look at the implementation of `DirectionalGaussianBlurFilterContents` and you wanted full test coverage it would be required.  In the description I mentioned that I split these off from another PR where I'm using them.  Looking ahead I know I'll need them.  I don't think it's ideal if all our tests that don't need hardware overrides are written one way and the ones that do are written against mocks.  I think the example presented should be usable for any future test writers.

**Comment by gaaclarke** [comment]
Jonah Matan and I met and chatted about this.  We agreed to move the tests to the concrete types and minimize the use of mocking to cases where it is unavoidable (such as upspecing the capabilities of the test hardware).  We also agreed to come up with a mechanism to downspec the capabilities in these concrete classes to aide testing.

File: impeller/renderer/testing/mocks.h [hunk 3]
```diff
@@ -164,5 +169,113 @@ class MockTexture : public Texture {
...
+ public:
+  MockRenderTargetAllocator(std::shared_ptr<Allocator> allocator)
+      : RenderTargetAllocator(allocator) {}
+  MOCK_METHOD(std::shared_ptr<Texture>,
+              CreateTexture,
+              (const TextureDescriptor& desc),
+              (override));
+};
+
+class MockCapabilities : public Capabilities {
```
Comment (jonahwilliams) [inline impeller/renderer/testing/mocks.h:197]: You don't need to MockCapabilities, there is a CapabilitiesBuilder/StandardCapabilities that lets you set whatever you want.

Comment (gaaclarke) [reply impeller/renderer/testing/mocks.h]: I just removed all this work for now.  They aren't used.  I figured it didn't hurt and may potentially save time in the future.

Comment (gaaclarke) [reply impeller/renderer/testing/mocks.h]: I'll probably end up landing most of this with the other blur anyways where it is actually used.  We'll see.

File: impeller/entity/contents/filters/directional_gaussian_blur_filter_contents_unittests.cc [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+  std::optional<Entity> result =
+      contents->GetEntity(*renderer, entity, /*coverage_hint=*/{});
+  EXPECT_TRUE(result.has_value());
+  if (result.has_value()) {
+    EXPECT_EQ(result.value().GetBlendMode(), BlendMode::kSourceOver);
+    std::optional<Rect> result_coverage = result.value().GetCoverage();
+    std::optional<Rect> contents_coverage = contents->GetCoverage(entity);
+    EXPECT_TRUE(result_coverage.has_value());
+    EXPECT_TRUE(contents_coverage.has_value());
+    if (result_coverage.has_value() && contents_coverage.has_value()) {
```
Comment (jonahwilliams) [inline impeller/entity/contents/filters/directional_gaussian_blur_filter_contents_unittests.cc:115]: Consider adding an Expect_rect_near macro

**Comment by gaaclarke** [comment]
> LGTM
> 
> nit: this is adding more testings only right? I would adjust PR title if so.

Nope, there is actually a coverage fix in there.

File: impeller/entity/contents/filters/directional_gaussian_blur_filter_contents_unittests.cc (hunk 4 above)
Comment (gaaclarke) [reply impeller/entity/contents/filters/directional_gaussian_blur_filter_contents_unittests.cc:115]: That'll problem happen if the new blur has better test coverage like i hope.

File: impeller/entity/contents/filters/directional_gaussian_blur_filter_contents.cc [hunk 5]
```diff
@@ -313,7 +313,9 @@ std::optional<Rect> DirectionalGaussianBlurFilterContents::GetFilterCoverage(
 
   auto transform = inputs[0]->GetTransform(entity) * effect_transform.Basis();
   auto transformed_blur_vector =
-      transform.TransformDirection(blur_direction_ * Radius{blur_sigma_}.radius)
+      transform
+          .TransformDirection(blur_direction_ *
+                              Radius{ScaleSigma(blur_sigma_)}.radius)
```
Comment (jonahwilliams) [inline impeller/entity/contents/filters/directional_gaussian_blur_filter_contents.cc:318]: Oh, missed this!

**Comment by jonahwilliams** [comment]
Oh, I missed that. Maybe "Ensure blur coverage sigma is also scaled..." or something like that.

**Pull request merged**